In [ ]:
# installing java, findspark, spark (initial setup)
%%bash
apt-get install openjdk-8-jdk-headless -qq > /dev/null
[ ! -e "$(basename spark-2.4.4-bin-hadoop2.7.tgz)" ] && wget  http://apache.osuosl.org/spark/spark-2.4.4/spark-2.4.4-bin-hadoop2.7.tgz
tar xf spark-2.4.4-bin-hadoop2.7.tgz
pip install -q findspark

## [**Start SPARK by setting the location it has to get installed**]

In [ ]:
import os
os.environ["JAVA_HOME"] = "/usr/lib/jvm/java-8-openjdk-amd64"
os.environ["SPARK_HOME"] = "/content/spark-2.4.4-bin-hadoop2.7"

In [ ]:
import findspark
findspark.init()
from pyspark.sql import SparkSession

# get a spark session.
spark = SparkSession.builder.master("local[*]").getOrCreate()

**Import Data from CSV and create DFs**

In [ ]:
! [ ! -e "$(basename Energy_Usage_2010.csv)" ] && wget  https://storage.googleapis.com/nishu_wsu/Energy_Usage_2010.csv

df = spark.read.csv('Jobsites_Dataset',header= True, inferSchema = True)


print(df.columns)


['COMMUNITY AREA NAME', 'CENSUS BLOCK', 'BUILDING TYPE', 'BUILDING_SUBTYPE', 'KWH JANUARY 2010', 'KWH FEBRUARY 2010', 'KWH MARCH 2010', 'KWH APRIL 2010', 'KWH MAY 2010', 'KWH JUNE 2010', 'KWH JULY 2010', 'KWH AUGUST 2010', 'KWH SEPTEMBER 2010', 'KWH OCTOBER 2010', 'KWH NOVEMBER 2010', 'KWH DECEMBER 2010', 'TOTAL KWH', 'ELECTRICITY ACCOUNTS', 'ZERO KWH ACCOUNTS', 'THERM JANUARY 2010', 'THERM FEBRUARY 2010', 'THERM MARCH 2010', 'TERM APRIL 2010', 'THERM MAY 2010', 'THERM JUNE 2010', 'THERM JULY 2010', 'THERM AUGUST 2010', 'THERM SEPTEMBER 2010', 'THERM OCTOBER 2010', 'THERM NOVEMBER 2010', 'THERM DECEMBER 2010', 'TOTAL THERMS', 'GAS ACCOUNTS', 'KWH TOTAL SQFT', 'THERMS TOTAL SQFT', 'KWH MEAN 2010', 'KWH STANDARD DEVIATION 2010', 'KWH MINIMUM 2010', 'KWH 1ST QUARTILE 2010', 'KWH 2ND QUARTILE 2010', 'KWH 3RD QUARTILE 2010', 'KWH MAXIMUM 2010', 'KWH SQFT MEAN 2010', 'KWH SQFT STANDARD DEVIATION 2010', 'KWH SQFT MINIMUM 2010', 'KWH SQFT 1ST QUARTILE 2010', 'KWH SQFT 2ND QUARTILE 2010', 'KWH 

**EDA on my dataset**

In [ ]:
type(df)

pyspark.sql.dataframe.DataFrame

In [ ]:
df.describe().show()

+-------+-------------------+--------------------+-------------+----------------+------------------+-----------------+------------------+-----------------+------------------+------------------+------------------+-----------------+------------------+-----------------+------------------+-----------------+------------------+--------------------+------------------+------------------+-------------------+------------------+------------------+------------------+------------------+-----------------+------------------+--------------------+------------------+-------------------+-------------------+------------------+------------------+-----------------+------------------+------------------+---------------------------+------------------+---------------------+---------------------+---------------------+-----------------+------------------+--------------------------------+---------------------+--------------------------+--------------------------+--------------------------+---------------------+---

In [ ]:
df.dtypes

[('COMMUNITY AREA NAME', 'string'),
 ('CENSUS BLOCK', 'bigint'),
 ('BUILDING TYPE', 'string'),
 ('BUILDING_SUBTYPE', 'string'),
 ('KWH JANUARY 2010', 'string'),
 ('KWH FEBRUARY 2010', 'string'),
 ('KWH MARCH 2010', 'string'),
 ('KWH APRIL 2010', 'string'),
 ('KWH MAY 2010', 'string'),
 ('KWH JUNE 2010', 'string'),
 ('KWH JULY 2010', 'string'),
 ('KWH AUGUST 2010', 'string'),
 ('KWH SEPTEMBER 2010', 'string'),
 ('KWH OCTOBER 2010', 'string'),
 ('KWH NOVEMBER 2010', 'string'),
 ('KWH DECEMBER 2010', 'string'),
 ('TOTAL KWH', 'string'),
 ('ELECTRICITY ACCOUNTS', 'string'),
 ('ZERO KWH ACCOUNTS', 'string'),
 ('THERM JANUARY 2010', 'string'),
 ('THERM FEBRUARY 2010', 'string'),
 ('THERM MARCH 2010', 'string'),
 ('TERM APRIL 2010', 'string'),
 ('THERM MAY 2010', 'string'),
 ('THERM JUNE 2010', 'string'),
 ('THERM JULY 2010', 'string'),
 ('THERM AUGUST 2010', 'string'),
 ('THERM SEPTEMBER 2010', 'string'),
 ('THERM OCTOBER 2010', 'string'),
 ('THERM NOVEMBER 2010', 'string'),
 ('THERM DECEMBE

**Print schema of the dataframes**

In [ ]:
df.printSchema()

root
 |-- COMMUNITY AREA NAME: string (nullable = true)
 |-- CENSUS BLOCK: long (nullable = true)
 |-- BUILDING TYPE: string (nullable = true)
 |-- BUILDING_SUBTYPE: string (nullable = true)
 |-- KWH JANUARY 2010: string (nullable = true)
 |-- KWH FEBRUARY 2010: string (nullable = true)
 |-- KWH MARCH 2010: string (nullable = true)
 |-- KWH APRIL 2010: string (nullable = true)
 |-- KWH MAY 2010: string (nullable = true)
 |-- KWH JUNE 2010: string (nullable = true)
 |-- KWH JULY 2010: string (nullable = true)
 |-- KWH AUGUST 2010: string (nullable = true)
 |-- KWH SEPTEMBER 2010: string (nullable = true)
 |-- KWH OCTOBER 2010: string (nullable = true)
 |-- KWH NOVEMBER 2010: string (nullable = true)
 |-- KWH DECEMBER 2010: string (nullable = true)
 |-- TOTAL KWH: string (nullable = true)
 |-- ELECTRICITY ACCOUNTS: string (nullable = true)
 |-- ZERO KWH ACCOUNTS: string (nullable = true)
 |-- THERM JANUARY 2010: string (nullable = true)
 |-- THERM FEBRUARY 2010: string (nullable = true)


**display the first 20 rows of the data frame**

In [ ]:
df.show()

+-------------------+---------------+-------------+----------------+----------------+-----------------+--------------+--------------+------------+-------------+-------------+---------------+------------------+----------------+-----------------+-----------------+---------+--------------------+-----------------+------------------+-------------------+----------------+---------------+--------------+---------------+---------------+-----------------+--------------------+------------------+-------------------+-------------------+------------+------------+--------------+-----------------+-------------+---------------------------+----------------+---------------------+---------------------+---------------------+----------------+------------------+--------------------------------+---------------------+--------------------------+--------------------------+--------------------------+---------------------+---------------+-----------------------------+------------------+-----------------------+-----

Fetching the distinct count of column values

In [ ]:
df.select('BUILDING TYPE').distinct().count()

4

In [ ]:
df.select('COMMUNITY AREA NAME').distinct().count()

77

**My dataset has two different types of energy classification. So Let's go ahead and create dataframes for differentiating the two energies.**


In [ ]:
#For Electrical Energy

electrical_energy = df.select(df.columns[:19])
display(electrical_energy)

DataFrame[COMMUNITY AREA NAME: string, CENSUS BLOCK: bigint, BUILDING TYPE: string, BUILDING_SUBTYPE: string, KWH JANUARY 2010: string, KWH FEBRUARY 2010: string, KWH MARCH 2010: string, KWH APRIL 2010: string, KWH MAY 2010: string, KWH JUNE 2010: string, KWH JULY 2010: string, KWH AUGUST 2010: string, KWH SEPTEMBER 2010: string, KWH OCTOBER 2010: string, KWH NOVEMBER 2010: string, KWH DECEMBER 2010: string, TOTAL KWH: string, ELECTRICITY ACCOUNTS: string, ZERO KWH ACCOUNTS: string]

In [ ]:
#For Thermal Energy

thermal_energy = df.select(df.columns[19:33])
display(thermal_energy)

DataFrame[THERM JANUARY 2010: string, THERM FEBRUARY 2010: string, THERM MARCH 2010: string, TERM APRIL 2010: string, THERM MAY 2010: string, THERM JUNE 2010: string, THERM JULY 2010: string, THERM AUGUST 2010: string, THERM SEPTEMBER 2010: string, THERM OCTOBER 2010: string, THERM NOVEMBER 2010: string, THERM DECEMBER 2010: string, TOTAL THERMS: string, GAS ACCOUNTS: string]

Dropping of a few columns from my newly created dataset.



In [ ]:
#Drop I

electrical_energy = electrical_energy.drop("CENSUS BLOCK")


In [ ]:
electrical_energy.show()

+-------------------+-------------+----------------+----------------+-----------------+--------------+--------------+------------+-------------+-------------+---------------+------------------+----------------+-----------------+-----------------+---------+--------------------+-----------------+
|COMMUNITY AREA NAME|BUILDING TYPE|BUILDING_SUBTYPE|KWH JANUARY 2010|KWH FEBRUARY 2010|KWH MARCH 2010|KWH APRIL 2010|KWH MAY 2010|KWH JUNE 2010|KWH JULY 2010|KWH AUGUST 2010|KWH SEPTEMBER 2010|KWH OCTOBER 2010|KWH NOVEMBER 2010|KWH DECEMBER 2010|TOTAL KWH|ELECTRICITY ACCOUNTS|ZERO KWH ACCOUNTS|
+-------------------+-------------+----------------+----------------+-----------------+--------------+--------------+------------+-------------+-------------+---------------+------------------+----------------+-----------------+-----------------+---------+--------------------+-----------------+
|     Archer Heights|  Residential|       Multi < 7|            null|             null|          null|          

We see that census block has been dropped and this was chosen to be dropped as it would not make any significant difference for my analysis.

In [ ]:
#Drop II
electrical_energy = electrical_energy.drop("BUILDING_SUBTYPE")
electrical_energy.show()
display(electrical_energy)

+-------------------+-------------+----------------+-----------------+--------------+--------------+------------+-------------+-------------+---------------+------------------+----------------+-----------------+-----------------+---------+--------------------+-----------------+
|COMMUNITY AREA NAME|BUILDING TYPE|KWH JANUARY 2010|KWH FEBRUARY 2010|KWH MARCH 2010|KWH APRIL 2010|KWH MAY 2010|KWH JUNE 2010|KWH JULY 2010|KWH AUGUST 2010|KWH SEPTEMBER 2010|KWH OCTOBER 2010|KWH NOVEMBER 2010|KWH DECEMBER 2010|TOTAL KWH|ELECTRICITY ACCOUNTS|ZERO KWH ACCOUNTS|
+-------------------+-------------+----------------+-----------------+--------------+--------------+------------+-------------+-------------+---------------+------------------+----------------+-----------------+-----------------+---------+--------------------+-----------------+
|     Archer Heights|  Residential|            null|             null|          null|          null|        null|         null|         null|           null|      

DataFrame[COMMUNITY AREA NAME: string, BUILDING TYPE: string, KWH JANUARY 2010: string, KWH FEBRUARY 2010: string, KWH MARCH 2010: string, KWH APRIL 2010: string, KWH MAY 2010: string, KWH JUNE 2010: string, KWH JULY 2010: string, KWH AUGUST 2010: string, KWH SEPTEMBER 2010: string, KWH OCTOBER 2010: string, KWH NOVEMBER 2010: string, KWH DECEMBER 2010: string, TOTAL KWH: string, ELECTRICITY ACCOUNTS: string, ZERO KWH ACCOUNTS: string]

We see that the building subtype has been dropped. This was because, the type of the building was my initial criteria.

# **Checking for the below conditions: **

1.   IsNull
2.   IsNaN







**Creating Alias of the dataframes subset**

In [ ]:
from pyspark.sql.functions import isnan, when, count, col

In [ ]:
electrical_energy.select([count(when(isnan(e) | col(e).isNull(), e)).alias(e) for e in electrical_energy.columns]).show()

+-------------------+-------------+----------------+-----------------+--------------+--------------+------------+-------------+-------------+---------------+------------------+----------------+-----------------+-----------------+---------+--------------------+-----------------+
|COMMUNITY AREA NAME|BUILDING TYPE|KWH JANUARY 2010|KWH FEBRUARY 2010|KWH MARCH 2010|KWH APRIL 2010|KWH MAY 2010|KWH JUNE 2010|KWH JULY 2010|KWH AUGUST 2010|KWH SEPTEMBER 2010|KWH OCTOBER 2010|KWH NOVEMBER 2010|KWH DECEMBER 2010|TOTAL KWH|ELECTRICITY ACCOUNTS|ZERO KWH ACCOUNTS|
+-------------------+-------------+----------------+-----------------+--------------+--------------+------------+-------------+-------------+---------------+------------------+----------------+-----------------+-----------------+---------+--------------------+-----------------+
|                  0|           77|             871|              871|           871|           871|         871|          871|          871|            871|      

In [ ]:
electrical_energy = electrical_energy.toDF('community_area_name',
 'building_type',
 'kwh_january_2010',
 'kwh_february_2010',
 'kwh_march_2010',
 'kwh_april_2010',
 'kwh_may_2010',
 'kwh_june_2010',
 'kwh_july_2010',
 'kwh_august_2010',
 'kwh_september_2010',
 'kwh_october_2010',
 'kwh_november_2010',
 'kwh_december_2010',
 'total_kwh_2010',
 'electrical_accounts',
 'zero_kwh_accounts')

Displaying the dataframe

In [ ]:
display(electrical_energy)

DataFrame[community_area_name: string, building_type: string, kwh_january_2010: string, kwh_february_2010: string, kwh_march_2010: string, kwh_april_2010: string, kwh_may_2010: string, kwh_june_2010: string, kwh_july_2010: string, kwh_august_2010: string, kwh_september_2010: string, kwh_october_2010: string, kwh_november_2010: string, kwh_december_2010: string, total_kwh_2010: string, electrical_accounts: string, zero_kwh_accounts: string]

Isnull check on building type column.

In [ ]:
from pyspark.sql.functions import col

electrical_energy = electrical_energy.where(col("building_type").isNotNull())

In [ ]:
electrical_energy.select([count(when(isnan(e) | col(e).isNull(), e)).alias(e) for e in electrical_energy.columns]).show()

+-------------------+-------------+----------------+-----------------+--------------+--------------+------------+-------------+-------------+---------------+------------------+----------------+-----------------+-----------------+--------------+-------------------+-----------------+
|community_area_name|building_type|kwh_january_2010|kwh_february_2010|kwh_march_2010|kwh_april_2010|kwh_may_2010|kwh_june_2010|kwh_july_2010|kwh_august_2010|kwh_september_2010|kwh_october_2010|kwh_november_2010|kwh_december_2010|total_kwh_2010|electrical_accounts|zero_kwh_accounts|
+-------------------+-------------+----------------+-----------------+--------------+--------------+------------+-------------+-------------+---------------+------------------+----------------+-----------------+-----------------+--------------+-------------------+-----------------+
|                  0|            0|             871|              871|           871|           871|         871|          871|          871|          

we can see there are several nan values in almost each of the columns, hence lets replace it with avg values.

Below is a function called fillna with average values in pyspark. This replaces nan values with avg of the entries

In [ ]:
import pyspark.sql.functions as fn

def fillna_with_avg(dataframe, exclude=set()):
    stats = dataframe.agg(*(
        fn.avg(e).alias(e) for e in dataframe.columns if e not in exclude
    ))
    return dataframe.na.fill(stats.first().asDict())

We apply this to columns given below to verify all the nan values are replaced with avg values

In [ ]:
electrical_energy = fillna_with_avg(electrical_energy, ['community_area_name','building_type'])

In [ ]:
electrical_energy.select([count(when(isnan(e) | col(e).isNull(), e)).alias(e) for e in electrical_energy.columns]).show()

+-------------------+-------------+----------------+-----------------+--------------+--------------+------------+-------------+-------------+---------------+------------------+----------------+-----------------+-----------------+--------------+-------------------+-----------------+
|community_area_name|building_type|kwh_january_2010|kwh_february_2010|kwh_march_2010|kwh_april_2010|kwh_may_2010|kwh_june_2010|kwh_july_2010|kwh_august_2010|kwh_september_2010|kwh_october_2010|kwh_november_2010|kwh_december_2010|total_kwh_2010|electrical_accounts|zero_kwh_accounts|
+-------------------+-------------+----------------+-----------------+--------------+--------------+------------+-------------+-------------+---------------+------------------+----------------+-----------------+-----------------+--------------+-------------------+-----------------+
|                  0|            0|               0|                0|             0|             0|           0|            0|            0|          

In [ ]:
electrical_energy.show(5)

+-------------------+-------------+-----------------+-----------------+---------------+-----------------+-----------------+-----------------+----------------+-----------------+------------------+-----------------+-----------------+-----------------+-----------------+-------------------+-----------------+
|community_area_name|building_type| kwh_january_2010|kwh_february_2010| kwh_march_2010|   kwh_april_2010|     kwh_may_2010|    kwh_june_2010|   kwh_july_2010|  kwh_august_2010|kwh_september_2010| kwh_october_2010|kwh_november_2010|kwh_december_2010|   total_kwh_2010|electrical_accounts|zero_kwh_accounts|
+-------------------+-------------+-----------------+-----------------+---------------+-----------------+-----------------+-----------------+----------------+-----------------+------------------+-----------------+-----------------+-----------------+-----------------+-------------------+-----------------+
|     Archer Heights|  Residential|422.3416870898885|499.3204854368932|520.5209960

Now for thermal energy subset dataframe


In [ ]:
from pyspark.sql.functions import isnan, when, count, col

In [ ]:
thermal_energy.select([count(when(isnan(t) | col(t).isNull(), t)).alias(t) for t in thermal_energy.columns]).show()

+------------------+-------------------+----------------+---------------+--------------+---------------+---------------+-----------------+--------------------+------------------+-------------------+-------------------+------------+------------+
|THERM JANUARY 2010|THERM FEBRUARY 2010|THERM MARCH 2010|TERM APRIL 2010|THERM MAY 2010|THERM JUNE 2010|THERM JULY 2010|THERM AUGUST 2010|THERM SEPTEMBER 2010|THERM OCTOBER 2010|THERM NOVEMBER 2010|THERM DECEMBER 2010|TOTAL THERMS|GAS ACCOUNTS|
+------------------+-------------------+----------------+---------------+--------------+---------------+---------------+-----------------+--------------------+------------------+-------------------+-------------------+------------+------------+
|              2230|               4232|            1482|           1575|          1857|           1767|           1820|             1908|                2282|              1722|               1559|               1544|        1296|        1296|
+------------------+

Data cleaning by changing the column headers from upper case to lower case and wiht no spaces in between

In [ ]:
thermal_energy = thermal_energy.toDF('therm_january_2010',
 'therm_february_2010',
 'therm_march_2010',
 'therm_april_2010',
 'therm_may_2010',
 'therm_june_2010',
 'therm_july_2010',
 'therm_august_2010',
 'therm_september_2010',
 'therm_october_2010',
 'therm_november_2010',
 'therm_december_2010',
 'total_therms_2010',
 'gas_accounts')

In [ ]:
thermal_energy.show()

+------------------+-------------------+----------------+----------------+--------------+---------------+---------------+-----------------+--------------------+------------------+-------------------+-------------------+-----------------+------------+
|therm_january_2010|therm_february_2010|therm_march_2010|therm_april_2010|therm_may_2010|therm_june_2010|therm_july_2010|therm_august_2010|therm_september_2010|therm_october_2010|therm_november_2010|therm_december_2010|total_therms_2010|gas_accounts|
+------------------+-------------------+----------------+----------------+--------------+---------------+---------------+-----------------+--------------------+------------------+-------------------+-------------------+-----------------+------------+
|             2,326|              2,131|           1,400|             620|           502|            224|            222|              187|                 197|               252|                744|              2,112|           10,917|          

Displaying the structure

In [ ]:
display(thermal_energy)

DataFrame[therm_january_2010: string, therm_february_2010: string, therm_march_2010: string, therm_april_2010: string, therm_may_2010: string, therm_june_2010: string, therm_july_2010: string, therm_august_2010: string, therm_september_2010: string, therm_october_2010: string, therm_november_2010: string, therm_december_2010: string, total_therms_2010: string, gas_accounts: string]

The common functions from electrical energy subset to thermal energy subset

In [ ]:
from pyspark.sql.functions import col

thermal_energy = thermal_energy.where(col("total_therms_2010").isNotNull())

In [ ]:
thermal_energy.select([count(when(isnan(t) | col(t).isNull(), t)).alias(t) for t in thermal_energy.columns]).show()

+------------------+-------------------+----------------+----------------+--------------+---------------+---------------+-----------------+--------------------+------------------+-------------------+-------------------+-----------------+------------+
|therm_january_2010|therm_february_2010|therm_march_2010|therm_april_2010|therm_may_2010|therm_june_2010|therm_july_2010|therm_august_2010|therm_september_2010|therm_october_2010|therm_november_2010|therm_december_2010|total_therms_2010|gas_accounts|
+------------------+-------------------+----------------+----------------+--------------+---------------+---------------+-----------------+--------------------+------------------+-------------------+-------------------+-----------------+------------+
|               934|               2936|             186|             279|           561|            471|            524|              612|                 986|               426|                263|                248|                0|          

The below function is filling all the na values with mean of the values.

In [ ]:
def fillna_with_mean(dataframe, exclude=set()):
    stats = dataframe.agg(*(
        fn.avg(t).alias(t) for t in dataframe.columns if t not in exclude
    ))
    return dataframe.na.fill(stats.first().asDict())

In [ ]:
thermal_energy = fillna_with_mean(thermal_energy) #,['gas_accounts']

In [ ]:
thermal_energy.select([count(when(isnan(t) | col(t).isNull(), t)).alias(t) for t in thermal_energy.columns]).show()

+------------------+-------------------+----------------+----------------+--------------+---------------+---------------+-----------------+--------------------+------------------+-------------------+-------------------+-----------------+------------+
|therm_january_2010|therm_february_2010|therm_march_2010|therm_april_2010|therm_may_2010|therm_june_2010|therm_july_2010|therm_august_2010|therm_september_2010|therm_october_2010|therm_november_2010|therm_december_2010|total_therms_2010|gas_accounts|
+------------------+-------------------+----------------+----------------+--------------+---------------+---------------+-----------------+--------------------+------------------+-------------------+-------------------+-----------------+------------+
|                 0|                  0|               0|               0|             0|              0|              0|                0|                   0|                 0|                  0|                  0|                0|          

In [ ]:
thermal_energy.show(5)

+------------------+-------------------+-----------------+------------------+-----------------+-----------------+-----------------+-----------------+--------------------+------------------+-------------------+-------------------+-----------------+------------+
|therm_january_2010|therm_february_2010| therm_march_2010|  therm_april_2010|   therm_may_2010|  therm_june_2010|  therm_july_2010|therm_august_2010|therm_september_2010|therm_october_2010|therm_november_2010|therm_december_2010|total_therms_2010|gas_accounts|
+------------------+-------------------+-----------------+------------------+-----------------+-----------------+-----------------+-----------------+--------------------+------------------+-------------------+-------------------+-----------------+------------+
|             2,326|              2,131|            1,400|               620|              502|              224|              222|              187|                 197|               252|                744|        

# Analysis begins

Checking for the max consumption of electrical energy in chicago 2010. Monthwise classification for the year 2010

This can be obtained by using the MAX function of the PYSPARK by converting it to RDD(rdd function)

In [ ]:
list_columns = electrical_energy.columns
list_month = ['january','february','march','april','may','june','july','august','september','october','november','december']
for i in range(len(list_month)):
  print("The max usage of Electricity for {} is {} joules".format(list_month[i], electrical_energy.select(list_columns[i+3]).rdd.max()[0]))
  print("\n")


The max usage of Electricity for january is 999 joules


The max usage of Electricity for february is 999 joules


The max usage of Electricity for march is 999 joules


The max usage of Electricity for april is 999 joules


The max usage of Electricity for may is 999 joules


The max usage of Electricity for june is 999 joules


The max usage of Electricity for july is 999 joules


The max usage of Electricity for august is 999 joules


The max usage of Electricity for september is 999 joules


The max usage of Electricity for october is 999 joules


The max usage of Electricity for november is 999 joules


The max usage of Electricity for december is 998,728 joules




From the obtained result, we can see that in December, Electric Energy Usage is Max in Chicago during 2010. This could be certainly due to the extreme weather conditions and demand for using heating conditions.

In [ ]:
list_columns = thermal_energy.columns
list_month = ['january','february','march','april','may','june','july','august','september','october','november','december']
for i in range(len(list_month)):
  print("The max usage of thermal for {} is {} joules".format(list_month[i], thermal_energy.select(list_columns[i+1]).rdd.min()[0]))
  print("\n")

The max usage of thermal for january is 1 joules


The max usage of thermal for february is 1 joules


The max usage of thermal for march is 1 joules


The max usage of thermal for april is 1 joules


The max usage of thermal for may is 1 joules


The max usage of thermal for june is 1 joules


The max usage of thermal for july is 1 joules


The max usage of thermal for august is 1 joules


The max usage of thermal for september is 1 joules


The max usage of thermal for october is 1 joules


The max usage of thermal for november is 1 joules


The max usage of thermal for december is 1,000 joules




The thermal energy consumption is MAX for the DECEMEBER month in Chicago during 2010. This is when the energy usage is proportional to the season and weather conditons.

Aggregate of the electrical energy in kwh in Chicago 2010

In [ ]:
electrical_energy.agg({"total_kwh_2010":"sum"}).collect()[0]

Row(sum(total_kwh_2010)=529139.3446327659)

The total is 529139 KWH

Aggregate of the thermal energy in joules in Chicago 2010

In [ ]:
thermal_energy.agg({"total_therms_2010":"sum"}).collect()[0]

Row(sum(total_therms_2010)=480459.0)

The total is 480459 joules.

Which Community area in Chicago consumed most Electrical Energy in 2010

In [ ]:
from pyspark.sql.types import IntegerType
electrical_energy = electrical_energy.withColumn("total_kwh_2010", electrical_energy["total_kwh_2010"].cast(IntegerType()))

We changed the datatype from string to int of the total kwh

In [ ]:
from pyspark.sql.functions import desc
electrical_energy.groupBy('community_area_name').sum('total_kwh_2010').orderBy(desc("SUM(total_kwh_2010)")).show()

+--------------------+-------------------+
| community_area_name|sum(total_kwh_2010)|
+--------------------+-------------------+
|           West Town|              21663|
|      West Englewood|              19732|
|        Logan Square|              18287|
|      Near West Side|              17640|
|           Englewood|              16856|
|     Lower West Side|              16368|
|              Austin|              16007|
|     Near North Side|              15624|
|            New City|              15359|
|      North Lawndale|              15027|
|       Humboldt Park|              14367|
|      Jefferson Park|              12096|
|      South Lawndale|              12090|
|         South Shore|              11909|
|        Chicago Lawn|              11533|
|      Belmont Cragin|              11329|
|            Lakeview|              11088|
|        Portage Park|              10691|
|Greater Grand Cro...|              10438|
|         Irving Park|              10140|
+----------

We can see from the results that West Town has the max usage of energy. This area is in the Loop county of chicago which has many high raised building , several malls and other commercial complexes.

In [ ]:
from pyspark.sql.functions import asc
electrical_energy.groupBy('community_area_name').sum('total_kwh_2010').orderBy(asc("SUM(total_kwh_2010)")).show()

+-------------------+-------------------+
|community_area_name|sum(total_kwh_2010)|
+-------------------+-------------------+
|             O'Hare|               null|
|        Forest Glen|               null|
|            Oakland|                504|
|          Hegewisch|                504|
|            Pullman|                504|
|      South Deering|                708|
|        Avalon Park|               1008|
|    Mount Greenwood|               1008|
|            Kenwood|               1008|
|           Burnside|               1087|
|        Fuller Park|               1097|
|            Ashburn|               1259|
|    Near South Side|               1512|
|        Edison Park|               2016|
|         North Park|               2176|
|          Montclare|               2385|
|    Calumet Heights|               2520|
|          West Lawn|               2520|
|            Dunning|               2520|
|               Loop|               2520|
+-------------------+-------------

We can see that from results that Loop has most consumption of energy.

Let's now do it for Thermal energy.

In [ ]:
from pyspark.sql.types import IntegerType
thermal_energy = thermal_energy.withColumn("total_therms_2010", thermal_energy["total_therms_2010"].cast(IntegerType()))

In [ ]:
from pyspark.sql.functions import desc
thermal_energy.groupBy('gas_accounts').sum('total_therms_2010').orderBy(desc("SUM(total_therms_2010)")).show()

+------------+----------------------+
|gas_accounts|sum(total_therms_2010)|
+------------+----------------------+
| Less than 4|                304920|
|           4|                 37235|
|           6|                 23543|
|           5|                 22300|
|           8|                 14079|
|           7|                 13004|
|           9|                 11293|
|          11|                  9121|
|          12|                  7518|
|          13|                  6636|
|          10|                  5937|
|          14|                  3497|
|          16|                  3491|
|          15|                  3393|
|          20|                  2600|
|          22|                  1911|
|          19|                  1667|
|          26|                  1520|
|          18|                  1502|
|          17|                  1452|
+------------+----------------------+
only showing top 20 rows



Buildings having gas accounts less than 4 have most therms used and this type of gas accounts have been used by commercial units mainly. This could be used by the malls and other commercial units in chicago.

Now let's see which building type has most power consumed.

In [ ]:
electrical_energy.groupBy('building_type').sum('total_kwh_2010').orderBy(desc("SUM(total_kwh_2010)")).show()

+-------------+-------------------+
|building_type|sum(total_kwh_2010)|
+-------------+-------------------+
|   Commercial|             355503|
|  Residential|             164785|
|   Industrial|               8064|
+-------------+-------------------+



We can see that the commercial building and complexes of the Chicago state consumbe more energy as compared to residential and indutrial units.

Now lets see the max crowd of chicago as per the community.

In [ ]:
chicago_total_energy = df.select('COMMUNITY AREA NAME','TOTAL POPULATION','TOTAL UNITS')
display(chicago_total_energy)

DataFrame[COMMUNITY AREA NAME: string, TOTAL POPULATION: string, TOTAL UNITS: string]

In [ ]:
from pyspark.sql.types import IntegerType
chicago_total_energy = chicago_total_energy.withColumn("TOTAL POPULATION", chicago_total_energy["TOTAL POPULATION"].cast(IntegerType()))

In [ ]:
from pyspark.sql.functions import desc
chicago_total_energy.groupBy('COMMUNITY AREA NAME').sum('TOTAL POPULATION').orderBy(desc("SUM(TOTAL POPULATION)")).show()

+-------------------+---------------------+
|COMMUNITY AREA NAME|sum(TOTAL POPULATION)|
+-------------------+---------------------+
|           Lakeview|               270475|
|          West Town|               233660|
|             Austin|               199894|
|       Lincoln Park|               175958|
|       Logan Square|               168840|
|             Uptown|               168247|
|         West Ridge|               163179|
|     South Lawndale|               159528|
|        Rogers Park|               155816|
|     Belmont Cragin|               155374|
|       Portage Park|               136365|
|          Edgewater|               134250|
|        Irving Park|               134043|
|    Near North Side|               130986|
|        South Shore|               130845|
|        Albany Park|               125348|
|      Humboldt Park|               121880|
|       Chicago Lawn|               110069|
|     Lincoln Square|               103016|
|           New City|           

We see that LakeView area of Chicago city has max population, its one of the 77 community areas of chicago and has many amenities. Hence, we see that this may lead to the high usage of enery in Chicago in 2010

# Now Lets see the Advantages and disadvantages between using PySpark and Pandas

# PySpark Over Pandas

### 1. In PySpark, we can execute parallel computations on multiple cores, kernals and execute multiple processes parallely. This would inturn increase the effiency. Whereas in Pandas, this type of parallel computation is not that efficient. In Pandas, dataframes are single server, in-memory. This would limit the size of the server memory.The Processing would be with a single server power.

### 2.PySpark is booming to become one of the best open source for data analysis procedures by providing platforms. For eg: this provides resilient distributed datasets a distributed collection of objects over pandas. Pandas has no rdd in them. Each dataset could be divided into logical partitions and could be used to compute different partitions of the cluster.

### 3.PySpark has many big data analysis tools and libraries for data analyts, which could be handy for analysing the data and learn them. In Pandas, the library is not offered to build models and analyse them.

# Pandas over PySpark

### 1.Pandas, a data analysis tool is  currently the most popular and mature open data source for data analysis tool. The libraries are highly optimised for performance.

### 2. Consider a complex where clause of sql in pyspark:In this case, Pandas was obviously fast in analysing the data and whereas, Spark struggled because of the complexity of the underlying data.

### 3. Joining the dataset (subset of the dataframe) would be another taxing task for the spark. This is common on a string column which would degrade the performance in spark

### 4.PySpark has no functions to determine the shape of the dataset.

### 5. PySpark is a framework for only large datasets on the other hand, pandas can process large data as well as smaller ones.